File: NUNYABU_Melchior/Fontend/EGA_FRONTEND/src/app/core/services/parametre.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Parametre {
  id?: number;
  cle: string;
  valeur: string;
  dateCreation?: string;
  dateModification?: string;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data?: T;
  errors?: { [key: string]: string };
}

@Injectable({
  providedIn: 'root'
})
export class ParametreService {
  private apiUrl = 'http://localhost:9090/api/parametres';

  constructor(private http: HttpClient) { }

  // Obtenir tous les paramètres
  getParametres(): Observable<ApiResponse<Parametre[]>> {
    return this.http.get<ApiResponse<Parametre[]>>(`${this.apiUrl}`);
  }

  // Obtenir un paramètre spécifique
  getParametre(cle: string): Observable<ApiResponse<Parametre>> {
    return this.http.get<ApiResponse<Parametre>>(`${this.apiUrl}/${cle}`);
  }

  // Créer/Mettre à jour un paramètre
  saveParametre(cle: string, valeur: string): Observable<ApiResponse<Parametre>> {
    return this.http.post<ApiResponse<Parametre>>(`${this.apiUrl}`, { cle, valeur });
  }

  // Supprimer un paramètre
  deleteParametre(cle: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${cle}`);
  }
}

File: NUNYABU_Melchior/Backend/ega/src/main/java/com/bank/ega/service/ParametreService.java
package com.bank.ega.service;

import com.bank.ega.entity.Parametre;
import com.bank.ega.entity.Utilisateur;
import com.bank.ega.exception.ClientNotFoundException;
import com.bank.ega.repository.ParametreRepository;
import com.bank.ega.repository.UtilisateurRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class ParametreService {

    private final ParametreRepository parametreRepository;
    private final UtilisateurRepository utilisateurRepository;

    public ParametreService(ParametreRepository parametreRepository, 
                           UtilisateurRepository utilisateurRepository) {
        this.parametreRepository = parametreRepository;
        this.utilisateurRepository = utilisateurRepository;
    }

    // Obtenir tous les paramètres d'un utilisateur
    public List<Parametre> getParametres(String username) {
        Utilisateur utilisateur = utilisateurRepository.findByUsername(username)
                .orElseThrow(() -> new ClientNotFoundException("Utilisateur non trouvé"));
        return parametreRepository.findByUtilisateur(utilisateur);
    }

    // Obtenir un paramètre spécifique
    public Optional<Parametre> getParametre(String username, String cle) {
        Utilisateur utilisateur = utilisateurRepository.findByUsername(username)
                .orElseThrow(() -> new ClientNotFoundException("Utilisateur non trouvé"));
        return parametreRepository.findByUtilisateurAndCle(utilisateur, cle);
    }

    // Créer ou mettre à jour un paramètre
    public Parametre saveParametre(String username, String cle, String valeur) {
        Utilisateur utilisateur = utilisateurRepository.findByUsername(username)
                .orElseThrow(() -> new ClientNotFoundException("Utilisateur non trouvé"));

        Optional<Parametre> existing = parametreRepository.findByUtilisateurAndCle(utilisateur, cle);

        Parametre parametre;
        if (existing.isPresent()) {
            parametre = existing.get();
            parametre.setValeur(valeur);
            parametre.setDateModification(LocalDateTime.now());
        } else {
            parametre = new Parametre();
            parametre.setUtilisateur(utilisateur);
            parametre.setCle(cle);
            parametre.setValeur(valeur);
            parametre.setDateCreation(LocalDateTime.now());
            parametre.setDateModification(LocalDateTime.now());
        }

        return parametreRepository.save(parametre);
    }

    // Supprimer un paramètre
    public void deleteParametre(String username, String cle) {
        Utilisateur utilisateur = utilisateurRepository.findByUsername(username)
                .orElseThrow(() -> new ClientNotFoundException("Utilisateur non trouvé"));

        Optional<Parametre> parametre = parametreRepository.findByUtilisateurAndCle(utilisateur, cle);
        if (parametre.isPresent()) {
            parametreRepository.delete(parametre.get());
        }
    }

    // Obtenir la valeur d'un paramètre avec défaut
    public String getParametreValeur(String username, String cle, String defaultValue) {
        return getParametre(username, cle)
                .map(Parametre::getValeur)
                .orElse(defaultValue);
    }
}
